wsclient/src/test/java/org/hpccsystems/ws/client/WSFileIOClientTest.java

```diff
@@ -67,45 +69,47 @@ public class WSFileIOClientTest extends BaseRemoteTest
     @Test
     public void copyFile() throws Exception
     {
-        Assume.assumeFalse("Test not valid on containerized HPCC environment", client.isTargetHPCCContainerized());
-        assumeTrue("Ignoring test 'copyFile' because HPCC-30117 is not fixed", HPCC_30117.equalsIgnoreCase("fixed"));
-
         String lzfile=System.currentTimeMillis() + "_csvtest.csv";
         String hpccfilename="temp::" + lzfile;
         client.createHPCCFile(lzfile, targetLZ, true);
         byte[] data = "Product,SKU,Color\r\nBike,1234,Blue\r\nCar,2345,Red\r\n".getBytes();
         client.writeHPCCFileData(data, lzfile, targetLZ, true, 0, 20);
         try
         {
-            ProgressResponseWrapper dfuspray=wsclient.getFileSprayClient().sprayVariable(
+            ProgressResponseWrapper dfuspray = wsclient.getFileSprayClient().sprayVariable(
                     new DelimitedDataOptions(),
                     wsclient.getFileSprayClient().fetchLocalDropZones().get(0),
                     lzfile,"~" + hpccfilename,"",thorClusterFileGroup,true,
                     HPCCFileSprayClient.SprayVariableFormat.DFUff_csv,
                     null, null, null, null, null, null, null);
-            Thread.sleep(1000);
-            int wait=60;
             if (dfuspray.getExceptions() != null
-                    && dfuspray.getExceptions().getException() != null
-                    && dfuspray.getExceptions().getException().size()>0)
+                && dfuspray.getExceptions().getException() != null
+                && dfuspray.getExceptions().getException().size()>0)
             {
                 fail(dfuspray.getExceptions().getException().get(0).getMessage());
             }
-            if (dfuspray.getSecsLeft()>0)
+
+            List<String> whiteListedStates = Arrays.asList( "queued", "started", "unknown", "finished", "monitoring");
+            int waitCount = 0;
+            int MAX_WAIT_COUNT = 60;
+
+            ProgressResponseWrapper dfuProgress = null;
+            do
             {
-                System.out.println("Still spraying, waiting 1 sec...");
-                for (int i=wait;i>0;i--)
+                dfuProgress = wsclient.getFileSprayClient().getDfuProgress(dfuspray.getWuid());
+                if (!whiteListedStates.contains(dfuProgress.getState())) {
```

Comment:
Since we're testing for a failure here, it feels more natural to me to seek for one of any blacklisted states instead. Is there something odd about the failure states that swayed you to seek for absence of whitelisted states?

Reply:
I generally go for whitelists because they are explicit and more likely to catch errors / future errors. IE: If a new state gets added the whitelist will treat it as an error but the blacklist would ignore it.

Reply:
Fair enough, but the original enum on the platform side doesn't seem to have changed in a while:
https://github.com/hpcc-systems/HPCC-Platform/blame/7552030a5bb03b63d4af0d37d454f8eab22feb9b/dali/dfu/dfuwu.hpp#L40
If that is the right enum, at this point I would consider it static.

With that in mind, it seems seeking explicitly for one of the known failed states will lead to less false positives.